InterfaceFortran/fortester.f90

```diff
@@ -27,184 +27,1483 @@
+double_1d_b = sparse_matmul( int_pnt_1d_a, int_pnt_1d_b, double_pnt_1d_a, weights )
+if ( norm1d(double_1d_b-double_2d_b(1,:)) > 1.d-11 ) then
+  write(*,*) "Mismatch in tsgEvaluateSparseHierarchicalFunctions: local grid test"
+  stop 1
+endif
+deallocate(points,double_2d_a,double_2d_b,weights,double_1d_b,rnd,pointsb,int_pnt_1d_a,int_pnt_1d_b,double_pnt_1d_a)
+
+
+! test reading a complex matrix
+call tsgMakeFourierGrid(gridID, 2, 1, 4, tsg_level)
+points => tsgGetPoints(gridID)
+i_a = tsgGetNumPoints(gridID)
+i_b = 13
+allocate(double_2d_a(1,i_a), double_2d_b(1,i_b), double_2d_c(i_b,2*i_a), &
+         double_1d_a(i_b),   double_1d_b(i_b),   double_1d_c(i_b), pointsb(2,i_b), double_1d_d(2*i_a))
+double_2d_a(1,:) = exp( -points(1,:)**2 - 2.d0 * points(2,:)**2 )
+call tsgLoadNeededPoints(gridID, double_2d_a)
+rnd     => rand(2,i_b)
+pointsb = -1.d0 + 2.d0 * rnd
+call tsgEvaluateBatch(gridID,pointsb,i_b,double_2d_b)
+call tsgEvaluateHierarchicalFunctions(gridID,pointsb,i_b,double_2d_c)
+call tsgGetHierarchicalCoefficientsStatic(gridID, double_1d_d)
+double_1d_a = Re_complex_matmul( double_2d_c, double_1d_d )
+
+allocate(dcmplx_1d_a(i_a),dcmplx_2d_c(i_b,i_a))
+call tsgEvaluateComplexHierarchicalFunctions(gridID,pointsb,i_b,dcmplx_2d_c)
+call tsgGetComplexHierarchicalCoefficientsStatic(gridID, dcmplx_1d_a)
+double_1d_b = real(matmul(dcmplx_2d_c,dcmplx_1d_a))
+
+call tsgGetComplexHierarchicalCoefficientsStatic(gridID, dcmplx_1d_a)
+dcmplx_pnt_1d_a => tsgGetComplexHierarchicalCoefficients(gridID)
+double_1d_c = real(matmul(dcmplx_2d_c,dcmplx_pnt_1d_a))
+
+if ( norm1d(double_1d_a - double_2d_b(1,:)) > 1.d-11 ) then
+  write(*,*) "Mismatch in tsgEvaluateSparseHierarchicalFunctions: fourier grid test 1"
+  stop 1
+endif
+if ( norm1d(double_1d_b - double_2d_b(1,:)) > 1.d-11 ) then
+  write(*,*) "Mismatch in tsgEvaluateSparseHierarchicalFunctions: fourier grid test 2"
+  stop 1
+endif
+if ( norm1d(double_1d_c - double_2d_b(1,:)) > 1.d-11 ) then
+  write(*,*) "Mismatch in tsgEvaluateSparseHierarchicalFunctions: fourier grid test 3"
+  stop 1
+endif
+deallocate(points,pointsb,double_2d_a,double_2d_b,double_2d_c,double_1d_a,double_1d_b,double_1d_c,double_1d_d,rnd,dcmplx_pnt_1d_a)
+
+
+! load coefficients
+! i_a = 13
+! allocate( double_2d_b(1,i_a), double_2d_c(1,i_a), pointsb(2,i_a) )
+! call tsgMakeLocalPolynomialGrid(gridID,    2, 1, 5, 1, tsg_semi_localp)
+! call tsgMakeLocalPolynomialGrid(gridID_II, 2, 1, 5, 1, tsg_semi_localp)
+! points  => tsgGetPoints(gridID);  i_b = tsgGetNumPoints(gridID)
+! rnd     => rand(2,i_a)
+! pointsb =  -1.d0 + 2.d0 * rnd
+! allocate( double_2d_a(1,i_b), double_1d_b(i_b), double_2d_d(i_b,i_b) )
+! double_2d_a(1,:) = exp( -points(1,:)**2 - 2.d0 * points(2,:)**2 )
+! call tsgLoadNeededPoints(gridID, double_2d_a)
+! call tsgEvaluateHierarchicalFunctions(gridID,points,i_b,double_2d_d)
+! double_1d_b = tsgSolveLeastSquares(double_2d_d,double_2d_a,0.d0)
+! call tsgSetHierarchicalCoefficients(gridID_II,double_1d_b)
```

Comment:
* So `tsgSetHierarchicalCoefficients()` is not tested right now? Here is the test:
```
call tsgMakeLocalPolynomialGrid(gridID,    2, 1, 5, 1, tsg_semi_localp)
call tsgMakeLocalPolynomialGrid(gridID_II, 2, 1, 5, 1, tsg_semi_localp)
...
call tsgLoadNeededPoints(gridID, double_2d_a)
double_1d_b = tsgGetHierarchicalCoefficients(gridID)
call tsgSetHierarchicalCoefficients(gridID_II,double_1d_b)
call tsgEvaluateBatch(gridID,pointsb,i_a,double_2d_b)
call tsgEvaluateBatch(gridID_II,pointsb,i_a,double_2d_c)
if ( norm2d(double_2d_b - double_2d_c) > 1.d-11 ) then
  ...
```
Basically, instead of using leas-squares, use another grid to compute the coefficients and see if you can properly copy-paste them from one grid to the next.

* remove the commented code